I18n dates 1395

tofumatt opened the pull request:
This adds support for moment and loading moment locales as part of our own (previously JSON, now JS) localisation files that we dynamically require depending on locale.

It's a bit of a change as we now have JS inside our locales. I've been at this several days and the way I got the locales in is a bit hairy but it works great in usage and avoids us loading all of moment's locales at once.

This adds moment info into the disco pane as well but seeing as disco pane doesn't use moment we might well be better off only adding them in for AMO. Let me know on that front, it's an easy change to make as it's all in the build script.

-----

While I was at it I added in the REASON I wanted moment; this tweak to the addon details (see the **last updated** bit:

<img width="320" alt="screenshot 2016-12-21 16 29 14" src="https://cloud.githubusercontent.com/assets/90871/21397574/4113f938-c79c-11e6-96fa-30da5c6875ec.png">
<img width="324" alt="screenshot 2016-12-21 16 29 02" src="https://cloud.githubusercontent.com/assets/90871/21397572/4110d208-c79c-11e6-86b8-501fabdd3119.png">

tofumatt commented:
Oh my, the locale changes make this PR wild... though I'm not sure the tests will work without them? I'll try to create a PR without them, hopefully it'll work and I'll build the locales separate.

mstriemer commented on bin/build-locales:
```diff
@@ -39,5 +40,23 @@ poFiles.forEach((pofile) => {
+  var defineLocale = null;
+  try {
+    const momentLocale = locale.replace('_', '-').toLowerCase();
+    defineLocale = new Function(
```
I might be worth commenting on what this does. At first I thought it would eval but it looks like it will actually end up wrapping the file in a function and printing that to file.

tofumatt replied:
Yeah, without doing this we end up writing `[object Object]`. Indeed though, it warrants an explanation.

mstriemer commented on bin/build-locales:
```diff
@@ -39,5 +40,23 @@ poFiles.forEach((pofile) => {
+    );
+  } catch (e) {
+    if (locale !== 'en_US') {
+      console.info(`No moment i18n available for ${locale}; consider adding one or creating a mapping.`);
```
This line is over 100 characters long. I'm guessing we don't lint the `bin` folder...

tofumatt replied:
I don't think so, but also I forgot if we had `dedent` here. But I bet we do, I'll add it.

tofumatt replied:
Actually maybe we don't. I'll just concat it here; the `bin/` folder is less pretty but I'll try not to make it a total garbage heap 👍

mstriemer replied:
I think you can use `oneLine` from common-tags. Pretty sure that got added back in recently.

mstriemer commented on src/amo/components/AddonMoreInfo.js:
```diff
@@ -32,7 +32,9 @@ export class AddonMoreInfoBase extends React.Component {
           <dt>{i18n.gettext('Last updated')}</dt>
-          <dd>{addon.last_updated}</dd>
+          <dd>
+            {i18n.moment(addon.last_updated).fromNow()} ({i18n.moment(addon.last_updated).format('MMMM Do YYYY')})
```
Should this be localised? I'm not sure if parens translate well and some languages might want something like "1 year ago on December 1, 2015"

mstriemer commented on src/core/i18n/translate.js:
```diff
@@ -14,6 +19,23 @@ export default function translate(options = {}) {
         super(props, context);
         this.i18n = context.i18n;
+
+        if (this.i18n) {
```
Can this move to server/base and client/base? Strictly speaking a component doesn't need to use `translate()` and it looks like this code might run more than once and keep configuring moment.

Changing this to `this.i18n && !this.i18n.momet` might fix that but moving it to where `i18n` is created probably makes the most sense.

Ryuno-Ki replied:
Typo: !this.i18n.mome*n*t

mstriemer commented on src/core/server/base.js:
```diff
@@ -182,17 +182,18 @@ function baseServer(routes, createStore, { appInstanceName = appName } = {}) {
             log.info(`Falling back to default lang: "${config.get('defaultLang')}".`);
           }
-          const i18n = new Jed(jedData);
+          const i18n = new Jed(i18nData);
```
Setting up moment here makes the most sense to me, provided there's nothing preventing that.

```js
function makeI18n(i18nData) {
  const i18n = new Jed(i18nData);
  moment.locale(/* yadda yadda */);
  i18n.moment = moment;
  return i18n;
}

const i18n = makeI18n(i18nData);

tofumatt commented:
Moved things into `client/base.js` and `server/base.js`, which is much nicer, agreed.

I added some comments and cleaned things up too.

Unrelated but some long lines in `base.js` were irking me so I fixed them but in a separate commit; if it's too noisy let me know and I'll put it elsewhere.

Ready for another r?

tofumatt commented:
Oh I also stopped adding moment locales to the disco pane. They're only added to AMO for now (it's hardcoded in the `build-locales` file but I'm just thinking we could probably add that as a config value, so maybe I'll do that in a few).

mstriemer commented on bin/build-locales:
```diff
@@ -39,5 +41,35 @@ poFiles.forEach((pofile) => {
+    try {
+      const momentLocale = locale.replace('_', '-').toLowerCase();
+      // We're using `new Function()` here to create a function out of the
+      // raw code in the file; if we require it here it will evaluate here
```
I think mentioning why this is being wrapped in a function more explicitly might be nice. Something like:

```js
// We're using `new Function()` here to create a function out of the
// raw code in the file; this function won't be executed but will be
// written out by `toSource()` so that it can be used later.

mstriemer commented on src/amo/components/AddonMoreInfo.js:
```diff
@@ -32,7 +32,16 @@ export class AddonMoreInfoBase extends React.Component {
+          <dd>
+            {i18n.sprintf(
+              // L10n: This will output, in English:
+              // "2 months ago (Dec 12 2016)"
```
That should probably be `Dec 12, 2016`.

mstriemer commented on src/amo/components/AddonMoreInfo.js:
```diff
@@ -32,7 +32,16 @@ export class AddonMoreInfoBase extends React.Component {
+              // "2 months ago (Dec 12 2016)"
+              i18n.gettext('%(timeFromNow)s (%(date)s)'), {
+                timeFromNow: i18n.moment(addon.last_updated).fromNow(),
+                date: i18n.moment(addon.last_updated).format('MMMM Do YYYY'),
```
I think we'll want to use the `ll` format for this. That will give `Dec 12, 2016` in English and `12 déc. 2016` in French.

tofumatt replied:
Oh, nice. Done.

mstriemer commented on src/core/i18n/utils.js:
```diff
@@ -173,3 +175,26 @@ export function getLanguage({ lang, acceptLanguage } = {}) {
+  // localised dates, times, etc.
+  if (i18n.options && typeof i18n.options._momentDefineLocale === 'function') {
+    i18n.options._momentDefineLocale();
+    moment.locale(makeMomentLocale(i18n.options.locale_data.messages[''].lang));
```
It looks like you can actually set a locale for an [instance of moment](http://momentjs.com/docs/#/i18n/instance-locale/). Maybe that would be better?

```js
const i18n = new _Jed(i18nData);
i18n.moment = moment();
if (i18n.options && typeof i18n.options._momentDefineLocale === 'function') {
  i18n.options._momentDefineLocale();
  i18n.moment.locale(makeMomentLocale(i18n.options.locale_data.messages[''].lang));
}
return i18n;

tofumatt replied:
I tried that, but frustratingly the docs don't seem to point out that the `moment()` returned by `moment()` is a `moment` object with `new Date()` set as the time. There isn't a way to make new moments like `moment(addon.last_updated)`. And I don't see an API to reset the date, so this doesn't actually work :-(

mstriemer replied:
That is really unfortunate. I guess we could attach a function that passes the args to moment(), sets the locale and returns it.

It might not be worth the complexity but if it works first try might be worth a shot. If it doesn't work right away I'd say leave it.

mstriemer commented on src/core/server/base.js:
```diff
@@ -259,16 +273,19 @@ export function runServer({ listen = true, app = appName } = {}) {
-            log.info(`🔥  Addons-frontend server is running [ENV:${env}] [APP:${app}] ` +
-                     `[isDevelopment:${isDevelopment}] [isDeployed:${isDeployed}] ` +
-                     `[apiHost:${config.get('apiHost')}] [apiPath:${config.get('apiPath')}]`);
-            log.info(`👁  Open your browser at http://${host}:${port} to view it.`);
```
We allow 100 character lines elsewhere so this line and some of the others could probably stay as they were.

tofumatt replied:
They are a lot easier to read under 80 and it still outputs the same:

<img width="802" alt="screenshot 2016-12-22 23 37 15" src="https://cloud.githubusercontent.com/assets/90871/21443779/9ccb7d60-c89f-11e6-8c61-d7e013e104e4.png">
vs
<img width="821" alt="screenshot 2016-12-22 23 37 07" src="https://cloud.githubusercontent.com/assets/90871/21443781/a0fad9da-c89f-11e6-9c27-a4ada411c4b9.png">

mstriemer commented on tests/client/core/i18n/test_utils.js:
```diff
@@ -387,4 +389,54 @@ describe('i18n utils', () => {
+      // This is annoying to work around because of the locale `require()`s
+      // and it only affects tests so it'd be nice to fix but doesn't break
+      // anything.
+      moment.locale('en');
```
If we were localising the moment instance you could check that `moment.locale()` is still `en` when setting the locale data to `fr` and remove this.
